File: libs/js/src/NodesInPathComparer.ts
import { Node } from './Node';
import { Path } from './Path';

export class NodesInPathComparer {
    private path: Path;

    private pairsNotInPath(node: Node) {
      let count = 0;
      for (let n of node.pairs) {
        if (this.path.contains(n.value)) {
          continue;
        }
        count++;
      }
      return count;
    }
  
    constructor(path: Path) {
      this.path = path;
    }
  
    compare(i: Node, j: Node) {
      let a = this.pairsNotInPath(i);
      let b = this.pairsNotInPath(j);
      if (a != b) {
        return a < b ? -1 : 1;
      }

      a = i.pairsCount();
      b = i.pairsCount();

      return a < b ? -1 : 1;
    }
  }

File: libs/dotnet/SquareSums/NodesInPathComparer.cs
using System.Collections.Generic;
using System.Linq;

namespace SquareSums
{
    internal class NodesComparer : Comparer<Node>
    {
        private readonly Path _path;

        public NodesComparer(Path p)
        {
            _path = p;
        }

        public override int Compare(Node? x, Node? y)
        {
            if (x == null || y == null)
            {
                return 0;
            }

            var a = PairsNotInPath(x, _path);
            var b = PairsNotInPath(y, _path);
            if (a != b)
            {
                return a.CompareTo(b);
            }

            a = x.PairsCount();
            b = y.PairsCount();
            
            if (a != b)
            {
                return a.CompareTo(b);
            }

            a = x.Value();
            b = y.Value();
            return b.CompareTo(a);
        }

        private static int PairsNotInPath(Node n, Path path)
        {
            return n.Pairs().Count(nn => !path.Contains(nn.Value()));
        }
    };
}